### PR TITLE
Feature: new firmware.py flasher

### PR DESCRIPTION
**Pull request by krautech**
# Firmware Flasher Script v0.0.1

image

A Python-based tool for managing and flashing firmware configurations for 3D printer setups. This script provides options to debug, select firmware versions, and handle configurations for CAN, USB, and DFU modes.

## Features

- **Interactive Menus:** Navigate through options with dynamic menus.
- **Debug Mode:** Enables detailed output for troubleshooting.
- **Firmware Selection:** Automatically retrieve and select firmware files based on device or user input.
- **Multi-mode Flashing:**
  - CAN: Supports flashing devices over CAN.
  - USB: Support for USB flashing.
  - DFU: Support for DFU flashing (coming soon).
- **High-Temperature Firmware Support:** Search specifically for high-temperature configurations.
- **Katapult Support:** Includes integration with the Katapult tool for CAN device management.

## Prerequisites

- Python 3.8+ installed on your system.
- Access to required tools such as `curl` and `tar` for firmware handling.

## Usage

Run the script with various command-line arguments to perform tasks.

### Command-Line Arguments

| Argument        | Description                                                                     | Example Usage                       |
|-----------------|---------------------------------------------------------------------------------|-------------------------------------|
| `-b`, `--branch` | Specify the branch to retrieve firmware from (default: `master`).                | `python3 firmware.py -b dev`       |
| `-D`, `--debug`  | Enable debug mode for detailed output.                                         | `python3 firmware.py -D`           |
| `-t`, `--type`   | Enable Katapult flash mode.                                                    | `python3 firmware.py -t`           |
| `-H`, `--high-temp` | Search for high-temperature firmware (`HT` directories).                   | `python3 firmware.py -H`           |
| `-a`, `--a` | Shows all firmware avaulable not just latest.               | `python3 firmware.py -a`           |
| `-k`, `--kseries` | Enable support for Creality K-Series printers.                                | `python3 firmware.py -k`           |
| `-f`, `--flash`  | Specify the flashing mode (`CAN`, `USB`, or `DFU`).                            | `python3 firmware.py -f CAN`       |
| `-d`, `--device` | Specify a device UUID for flashing.                                            | `python3 firmware.py -d <UUID>`    |

#### Flash Firmware via CAN
```bash
python3 firmware.py -f CAN
```

#### Enable Debug Mode
```bash
python3 firmware.py -D
```

#### Flash High-Temperature Firmware
```bash
python3 firmware.py -H -f CAN
```

## Menus

The script provides an interactive menu-based interface. Key menus include:

- **Main Menu:** Choose between CAN, USB, and DFU modes.
- **CAN Menu:** Options to find and flash CAN devices.
- **Firmware Menu:** List available firmware files and select one for flashing.
- **Confirmation Menu:** Confirm selected firmware and device before flashing.

## Debugging

Use the `-d` flag to enable debug mode, which provides detailed logs of operations.

### CODE REVIEW

File: scripts/README.md [hunk 1]
```diff
@@ -1,7 +1,81 @@
...
+```bash
+python3 firmware.py -l -D <device>
+```
+
+## Menus
+
+The script provides an interactive menu-based interface. Key menus include:
+
+- **Main Menu:** Choose between CAN, USB, and DFU modes.
+- **CAN Menu:** Options to find and flash CAN devices.
```
Comment (Jomik) [inline scripts/README.md:67]: Is there USB Menu?

File: scripts/firmware.py [hunk 2]
```diff
@@ -0,0 +1,1407 @@
...
+                            if isinstance(exclude_pattern, list):
+                                # Skip files if they match any pattern in the list
+                                if any(
+                                    fnmatch.fnmatch(file, pattern)
+                                    for pattern in exclude_pattern
+                                ):
+                                    continue
+                            elif fnmatch.fnmatch(file, exclude_pattern):
+                                continue
+                        firmware_files.append((subdirectory, file))
```
Comment (Jomik) [inline scripts/firmware.py:306]: Try to flatten this. Use "early return" or in this case an early continue.
`if not ... then return`

File: scripts/firmware.py [hunk 3]
```diff
@@ -0,0 +1,1407 @@
...
+
+        return sorted(firmware_files)  # Sort the results
+
+    def select_latest(self, firmware_files: List[Tuple[str, str]], type: str):
```
Comment (Jomik) [inline scripts/firmware.py:310]:
```suggestion
    def select_latest(self, firmware_files: List[Thingy?], type: str):

```
Can we make `Thingy?` a thing? You can use a `NamedTuple` or `TypedDict`.
I have no idea of what either str is.

File: scripts/firmware.py [hunk 4]
```diff
@@ -0,0 +1,1407 @@
...
+
+            # Update self.dir_path only once
+            self.dir_path = base_path
+            firmware_files = self.find_firmware_files(
+                self.dir_path, search_pattern, exclude_pattern, self.high_temp
+            )
+            if self.latest:
+                self.select_latest(firmware_files, type)
+            else:
+                self.display_firmware_menu(firmware_files, type)
```
Comment (Jomik) [inline scripts/firmware.py:439]: This could an isolated function.

**Review by Jomik — COMMENTED** [review]
Read through a bit - it's what I have time for right now.
If you want to improve maintainability/readability, I'd suggest reading through from top to bottom and see if you can follow the functions, assuming you know nothing.
There's a lot of logic crammed into rather long functions. It's better to make many smaller functions where you have logical separations/blocks.

File: scripts/firmware.py (hunk 3 above)
Comment (krautech) [reply scripts/firmware.py:642]: its (dir, filename) if you look 1 line above :D

File: scripts/firmware.py (hunk 2 above)
Comment (krautech) [reply scripts/firmware.py]: Should be resolved https://github.com/Cartographer3D/cartographer-klipper/pull/136/commits/c9f88adb07ba2677596d5467cee39e7763b878b7

File: scripts/README.md (hunk 1 above)
Comment (krautech) [reply scripts/README.md:63]: Yes

File: scripts/firmware.py (hunk 3 above)
Comment (krautech) [reply scripts/firmware.py]: https://github.com/Cartographer3D/cartographer-klipper/pull/136/commits/e1e140190919043b97681bd7377db21920c23707

File: scripts/firmware.py [hunk 5]
```diff
@@ -0,0 +1,2097 @@
...
+import re
+import subprocess
+import argparse
+import shutil
+import tempfile
+import fnmatch
+import platform
+import time
+
+from enum import Enum  # type: ignore
```
Comment (Jomik) [inline scripts/firmware.py]:
```suggestion
from enum import Enum
```
These are bad and shouldn't be used. Minimum should be `pyright: ignore[...]` as we do in scanner.py
Why is this necessary?

File: scripts/firmware.py [hunk 6]
```diff
@@ -0,0 +1,2097 @@
...
+    type: bool = False
+    high_temp: bool = False
+    all: bool = False
+    kseries: bool = False
+    device: Optional[str] = None
+    flash: Optional[str] = None
+
+
+class Version(NamedTuple):
+    parts: Tuple[int, ...]
```
Comment (Jomik) [inline scripts/firmware.py]: A named tuple with one thing? 😂 
Also, couldn't the version be `-beta.1`?
So a part could be a string?
I'd probably make it 
```suggestion
class Version(TypedDict):
    major: int
    minor: int
    patch: int
    prerelease: str
    build: str
```

File: scripts/firmware.py [hunk 7]
```diff
@@ -0,0 +1,2097 @@
...
+
+    @classmethod
+    def from_string(cls, version: str) -> "Version":
+        return cls(tuple(int(part) for part in version.split(".")))
+
+
+class FirmwareFile(NamedTuple):
+    subdirectory: str
+    filename: str
```
Comment (Jomik) [inline scripts/firmware.py]: Or `TypedDict`, depends if you want a tuple or a dict.

File: scripts/firmware.py [hunk 8]
```diff
@@ -0,0 +1,2097 @@
...
+
+
+class Utils:
```
Comment (Jomik) [inline scripts/firmware.py]: Never been a fan of "utils", but acceptable :)

File: scripts/firmware.py [hunk 9]
```diff
@@ -0,0 +1,2097 @@
...
+        # Center the mode string
+        mode = mode.center(PAGE_WIDTH)
+        print(Utils.colored_text(mode, Color.RED))
+
+
+class Menu:
+    title: str
+    menu_items: Dict[int, Union["Menu.MenuItem", "Menu.Separator"]]
+
+    class MenuItem:
```
Comment (Jomik) [inline scripts/firmware.py]:
```suggestion
    class Item:
```
I'd just call it `Item`, adding `Menu` seems redundant.

File: scripts/firmware.py [hunk 10]
```diff
@@ -0,0 +1,2097 @@
...
+                "\nPress Enter to return to the main menu...", Color.YELLOW
+            )
+        )
+        self.firmware.main_menu()
+
+
+class Firmware:
+    can: "Can"
+    usb: "Usb"
+    dfu: "Dfu"
```
Comment (Jomik) [inline scripts/firmware.py]: Should this be an enum?
```suggestion
class FlashMethod(StrEnum):
	CAN: auto()
	USB: auto()
	DFU: auto()

```

File: scripts/firmware.py [hunk 11]
```diff
@@ -0,0 +1,2097 @@
...
+        self.high_temp: bool = high_temp
+        self.flash: Optional[str] = flash
+        self.kseries: bool = kseries
+        self.all: bool = all
+        self.device: Optional[str] = device
+        self.can = Can(self, debug=self.debug, ftype=self.ftype)
+        self.usb = Usb(self, debug=self.debug, ftype=self.ftype)
+        self.dfu = Dfu(
+            self, debug=self.debug, ftype=self.ftype
+        )  # Pass Firmware instance to CAN
```
Comment (Jomik) [inline scripts/firmware.py]: This is string type on the class instance, now it's some other thing.
What is this?

File: scripts/firmware.py [hunk 12]
```diff
@@ -0,0 +1,2097 @@
...
+
+        # Retrieve the appropriate handler and call it if valid
+        handler = menu_handlers.get(type)
+        if handler:
+            handler()  # Call the appropriate menu method
+        else:
+            Utils.error_msg("You have not selected a valid firmware file.")
+
+    # Show a list of available firmware
+    def firmware_menu(self, type: str):
```
Comment (Jomik) [inline scripts/firmware.py]:
```suggestion
    def firmware_menu(self, type: FlashMethod):

```

File: scripts/firmware.py [hunk 13]
```diff
@@ -0,0 +1,2097 @@
...
+            firmware_files = self.find_firmware_files(
+                self.dir_path, search_pattern, exclude_pattern, self.high_temp
+            )
+            if not self.all:
+                self.select_latest(firmware_files, type)
+            else:
+                self.display_firmware_menu(firmware_files, type)
+
+    # Confirm the user wants to flash the correct device & file
+    def confirm(self, type: str):
```
Comment (Jomik) [inline scripts/firmware.py]:
```suggestion
    def confirm(self, type: FlashMethod):

```

**Review by Jomik — DISMISSED** [review]

File: scripts/firmware.py (hunk 9 above)
Comment (krautech) [reply scripts/firmware.py]: Makes sense

File: scripts/firmware.py (hunk 7 above)
Comment (krautech) [reply scripts/firmware.py]: Should be fine as a tuple, Doesnt need to be mutable I dont think at this time.

File: scripts/firmware.py (hunk 6 above)
Comment (krautech) [reply scripts/firmware.py]: Done

File: scripts/firmware.py (hunk 5 above)
Comment (krautech) [reply scripts/firmware.py]: Im not actually sure what caused this? Its removed.

File: scripts/firmware.py [hunk 14]
```diff
@@ -0,0 +1,2144 @@
...
+    debug: bool = False
+    type: bool = False
+    high_temp: bool = False
+    all: bool = False
+    kseries: bool = False
+    device: Optional[str] = None
+    flash: Optional[str] = None
+
+
+class Version(TypedDict):
```
Comment (Jomik) [inline scripts/firmware.py:67]: I don't think this is used anywhere?

**Review by Jomik — CHANGES_REQUESTED** [review]
Probably fine, other than the fucked commit history :D

File: scripts/firmware.py (hunk 14 above)
Comment (krautech) [reply scripts/firmware.py:67]: Version? or flash?

Comment (krautech) [reply scripts/firmware.py:67]: When i remove the flash line i get type errors elsewhere

Comment (Jomik) [reply scripts/firmware.py:67]: Version, woops.
